Classes/EventStore/EventStreamFilterInterface.php

```diff
@@ -13,22 +13,35 @@
 
 interface EventStreamFilterInterface
 {
-    public function getStreamName(): string;
-
-    public function hasStreamName(): bool;
-
-    public function getStreamNamePrefix(): string;
+    /**
+     * string representing the exact stream name
+     */
+    const FILTER_STREAM_NAME = 'streamName';
 
-    public function hasStreamNamePrefix(): bool;
+    /**
+     * string representing a stream name prefix. Any stream starting with the same string matches this filter
+     */
+    const FILTER_STREAM_NAME_PREFIX = 'streamNamePrefix';
 
     /**
-     * @return string[] in the format ['Bounded.Context:SomeEvent', 'Bounded.Context:SomeOtherEvent', ...]
+     * array of strings in the format ['Bounded.Context:SomeEvent', 'Bounded.Context:SomeOtherEvent', ...]
      */
-    public function getEventTypes(): array;
+    const FILTER_EVENT_TYPES = 'eventTypes';
 
-    public function hasEventTypes(): bool;
+    /**
+     * integer with the minimum sequence number to be matched
+     */
+    const FILTER_MINIMUM_SEQUENCE_NUMBER = 'minimumSequenceNumber';
 
-    public function getMinimumSequenceNumber(): int;
+    /**
+     * string representing the correlationId Metadata that has to match
+     */
+    const FILTER_CORRELATION_ID = 'correlationId';
 
-    public function hasMinimumSequenceNumber(): bool;
+    /**
+     * This method is expected to return an array where the keys can be one or more of the FILTER_* constants
+     *
+     * @return array
+     */
+    public function getFilterValues(): array;
```

Comment:
Changed this so that we don't have to implement/update loads of empty methods when creating new dedicated filters

Reply:
I'm unsure about this. I see the hassle of the empty methods, but with this we forfeit all of the type safety and implementation clarity for some additional flexibility/conciseness. Also, the actual filtering method now is much more technical due to the `array_key_exists` checks with the constants.

Can we maybe just bring down the hassle with an `AllEventsStreamFilter` implementation, so that concrete filters only need to extend that and override the specifc methods that they want to filter on?

Reply:
Just to clarify: I see the problem with the current implementation not so much in the hassle of implementing methods but in the lack of extensibility.
To implement a new filter we currently have to *change the interface* which will break *every implementation*.

> we forfeit all of the type safety

Yes, there's *slightly* reduced type safety but I considered it a pragmatic solution as this won't be used in highlevel userland code anyways.

> the actual filtering method now is much more technical 

is that a problem? That happens only within the EventStore implementation, which is very technical anyways.

> Can we maybe just bring down the hassle with an AllEventsStreamFilter

I would really like to avoid abstract base classes to solve this and a QOM/DSL would be overkill IMO.
What I could imagine is a purely interface-based approach like

```php
interface FiltersStreamNameInterface
{
  public function getStreamName(): string
}
interface FiltersStreamNamePrefixInterface
{
  public function getStreamNamePrefix(): string
}
// ...
```

but is that really worth it?

Reply:
> To implement a new filter we currently have to change the interface which will break every implementation.

Fair point, but my take would be, that there is an upper boundary on what the EventFilter interface should contain, because there's only so much that makes (generically) sense to filter events by. Anything on top can still be solved with an untyped "additionalFilterProperties(): array" method, but the basic filters should be expressive and type safe.

> is that a problem? That happens only within the EventStore implementation, which is very technical anyways.

You're right, I just prefer to to read something like `if ($foo->hasSomething())` over `if (array_key_exists($foo, FooClass::SOMETHING))` any time, even inside technical code.

> I would really like to avoid abstract base classes

Doesn't need to be abstract even. Think of it like this: Any (custom) filter is a specialization of an `AllEventsStreamFilter`. It is a real **is-a** relationship, so inheritance is a plausible modelling pattern.

> What I could imagine is a purely interface-based approach like

TBH I would even prefer that over a single array typed filter specification interface. But it would lead to lots of `if ($foo instanceof Bar)` checks - PHP finally needs some real pattern matching soon :/